impeller/entity/contents/filters/gaussian_blur_filter_contents.cc

```diff
@@ -284,8 +287,19 @@ std::optional<Entity> GaussianBlurFilterContents::RenderFilter(
                                 entity.GetClipDepth());  // No blur to render.
   }
 
+  if (input_snapshot.value().texture->NeedsMipmapGeneration()) {
```

Comment:
Thinking about this some more, I think in order to make this completely safe, we'd actually need to insert this blit pass just before we apply any backdrop filter.

A good place to do this would be right after we end the parent pass in preparation for reading: https://github.com/flutter/engine/blob/d13479968b8f2c1ac358a12b5d89dd76f7f0aa88/impeller/entity/entity_pass.cc#L551

Reply:
Ah also, when a layer texture is drawn, it may use mip sampling as well. We may apply an image filter to it.

So additionally you'll need to check `NeedsMipmapGeneration()` for `subpass_texture` and insert a blit pass if necessary just before returning here: https://github.com/flutter/engine/blob/d13479968b8f2c1ac358a12b5d89dd76f7f0aa88/impeller/entity/entity_pass.cc#L664-L674

Reply:
If both of these cases aren't covered then it's possible for the draw to fail, since sampling from a texture marked as having an invalid mipmap is considered an error if the sampler has mip sampling turned on.

Reply:
Oh, by "blit pass" he meant the generation of the mipmaps.

Reply:
@bdero I'm not keen on having to remember to do this in 2 places.  Could we potentially do something where like on the render pass when someone submits a command buffer that renders to that render target it will automatically insert the mip map generation?  That way we could capture both use cases with one mechanism.  Ideally it would be something we could setup in the same code when we specify that we want mips.

```c++
auto render_target = MakeRenderTarget(/*mip_count=*/4);
render_pass_->AddRenderHook(render_target, [] {
  render_target->GetTexture()->GenerateMipmap();
});
```

Reply:
"Rendering" happens in two stages in the HAL. First, passes are "encoded" into a command buffer -- order of encoding matters. Then, the command buffer is "submitted" -- the order in which CommandBuffers are submitted is also matters.

A good singular place to perform the blit if necessary is in `InlinePassContext::EndPass`. Currently we're encoding the pass and submitting the command buffer with `CommandBuffer::EncodeAndSubmit`:
https://github.com/flutter/engine/blob/6cb256fc49857571209a03b31e9b8c38843a2b0f/impeller/entity/inline_pass_context.cc#L59-L64

Instead, we could just generate the mipmaps by appending a pass to the same command buffer before we submit:
```cpp
  if (command_buffer_) {
    if (!pass_->EncodeCommands()) {
      VALIDATION_LOG << "Failed to encode the render pass.";
      return false;
    }
    
    auto pass_texture = pass_target_.GetRenderTarget().GetRenderTargetTexture();
    if (pass_texture->NeedsMipmapGeneration()) {
      auto blit_pass = command_buffer_->CreateBlitPass();
      blit_pass->GenerateMipmap(pass_texture);
      blit_pass->EncodeCommands(context_->GetResourceAllocator());
    }

    if (!command_buffer_->SubmitCommands()) {
      VALIDATION_LOG << "Failed to submit command buffer.";
    }
  }
```

Reply:
I was looking at InlinePassContext, is that going to catch your imagefilter case though too?

Reply:
My thought process was:
* I thought the hook idea was nifty
* Figured it could be a post-encoding hook on RenderPass, like `render_pass->AddPostEncodeHook([](CommandBuffer& command_buffer) {})`
* Remembered that RenderPasses can only be encoded once, so perhaps it can just be a "CompletionCallback" for `RenderPass::EncodeCommands()`, just like we do for CommandBuffer submissions.
* Remembered that encoding is always synchronous, so this isn't necessary.

Reply:
> I was looking at InlinePassContext, is that going to catch your imagefilter case though too?

Ugh, I missed this comment when writing the one above. It's 2024 and we ain't even ajax yet. 🙄

Yeah this should cover all cases, since the `RenderPass` that writes to the `EntityPass` texture always needs to end before something reads from it -- whether that be a backdrop filter read or drawing to the parent pass.